scripts/gen-live-site/main.go

```diff
@@ -0,0 +1,271 @@
+			return fmt.Errorf("unknown file type: %s: %s", mode, srcfile)
+		}
+
+		switch ext {
+		case ".html":
+			skippedFiles = append(skippedFiles, filename)
+			return nil
+		case ".js", ".css":
+			shortSha, err := hashFile(srcfile)
+			if err != nil {
+				return err
+			}
+			basename := strings.TrimSuffix(filepath.Base(filename), ext)
+			target := basename + "." + shortSha + ext
+			destfile = filepath.Join(filepath.Dir(destfile), target)
+			if _, ok := renamedFiles[filename]; ok {
+				//nolint:goerr113 // dynamic errors in package main is ok
+				return fmt.Errorf("duplicate filename: %s", srcfile)
+			}
+			renamedFiles[filename] = target
+		}
+		return copyFile(srcfile, destfile)
+	})
+	if err != nil {
+		return nil, nil, err
+	}
+	return skippedFiles, renamedFiles, nil
+}
+
+func copyHTMLFiles(src, dest, domain string, skippedFiles []string, renamedFiles map[string]string) error {
+	for _, filename := range skippedFiles {
+		in, out, err := openInOut(filepath.Join(src, filename), filepath.Join(dest, filename))
+		if err != nil {
+			return err
+		}
+		defer in.Close() //nolint:errcheck // don't care about close failing on read-only files
+		err = updateHTMLFile(out, in, filename, domain, renamedFiles)
+		if err != nil {
+			out.Close() //nolint:errcheck,gosec // we're returning the more important error
+			return err
+		}
+		if err := out.Close(); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// hashFile returns a short hash of the contents of filename. The short hash is
+// 32 bits, or 8 chars[0-9a-f] and with 100 file changes in a year (cache
+// expiry is one year) has a collision probability of less than 0.0000000005%.
+func hashFile(filename string) (string, error) {
+	f, err := os.Open(filename)
+	if err != nil {
+		return "", err
+	}
+	defer f.Close() //nolint:errcheck // don't care about close failing on read-only files
+	h := sha256.New()
+	if _, err := io.Copy(h, f); err != nil {
+		return "", err
+	}
+	sha := h.Sum(nil)
+	return hex.EncodeToString(sha[:4]), nil
+}
+
+func openInOut(src, dest string) (io.ReadCloser, io.WriteCloser, error) {
+	in, err := os.Open(src)
+	if err != nil {
+		return nil, nil, err
+	}
+	out, err := os.Create(dest)
+	if err != nil {
+		in.Close() //nolint:errcheck,gosec // we're returning the more important error
+		return nil, nil, err
+	}
+	return in, out, nil
+}
+
+func copyFile(src, dest string) error {
+	in, out, err := openInOut(src, dest)
+	if err != nil {
+		return err
+	}
+	defer in.Close() //nolint:errcheck // don't care about close failing on read-only files
+	_, err = io.Copy(out, in)
+	if err != nil {
+		out.Close() //nolint:errcheck,gosec // we're returning the more important error
+		return err
+	}
+	return out.Close()
+}
+
+var (
+	subdomainRE = regexp.MustCompile(`(href|value)="/(discord|docs|learn|play)`)
+	jscssRefRE  = regexp.MustCompile(`(href|src)="(.*\.(?:css|js))"`)
+	importmapRE = regexp.MustCompile(`"(.*\.js)": "(.*\.js)"`)
```

Comment:
seeing as this may become a slightly more robust/serious CLI now, I wonder if these REs should come via flags as well? probably not, just worthy of a thought

Reply:
I'm not sure - they do not stand alone - the replacement part uses them but constructs the replacement string programmatically. There does appear to be an opportunity to refactor the replacement stuff but I have not come up with a way that is clear, so I left the duplication of the 3 replacements in place. If I could make that generic and clear, I would consider making the regexps user-pluggable but I'll leave it for now.

Reply:
ok, agreed.